src/storage/src/upsert/types.rs

```diff
@@ -364,124 +656,126 @@ impl<O: Default> StateValue<O> {
     /// then mutate it to its `Value` state, so the `upsert` operator can use it.
     #[allow(clippy::as_conversions)]
-    pub fn ensure_decoded(&mut self, bincode_opts: BincodeOpts) {
+    pub fn ensure_decoded(&mut self, bincode_opts: BincodeOpts) -> Result<(), anyhow::Error> {
         match self {
-            StateValue::Snapshotting(snapshotting) => {
-                match snapshotting.diff_sum.0 {
+            StateValue::Consolidating(consolidating) => {
+                match consolidating.diff_sum.0 {
                     1 => {
-                        let len = usize::try_from(snapshotting.len_sum.0)
-                            .map_err(|_| {
-                                format!(
-                                    "len_sum can't be made into a usize, state: {}",
-                                    snapshotting
-                                )
-                            })
-                            .expect("invalid upsert state");
-                        let value = &snapshotting
-                            .value_xor
-                            .get(..len)
-                            .ok_or_else(|| {
-                                format!(
-                                    "value_xor is not the same length ({}) as len ({}), state: {}",
-                                    snapshotting.value_xor.len(),
-                                    len,
-                                    snapshotting
-                                )
-                            })
-                            .expect("invalid upsert state");
+                        let len = usize::try_from(consolidating.len_sum.0).map_err(|_| {
+                            anyhow::anyhow!(
+                                "len_sum can't be made into a usize, state: {}",
+                                consolidating
+                            )
+                        })?;
+                        let value = &consolidating.value_xor.get(..len).ok_or_else(|| {
+                            anyhow::anyhow!(
+                                "value_xor is not the same length ({}) as len ({}), state: {}",
+                                consolidating.value_xor.len(),
+                                len,
+                                consolidating
+                            )
+                        })?;
                         // Truncation is fine (using `as`) as this is just a checksum
-                        assert_eq!(
-                            snapshotting.checksum_sum.0,
-                            // Hash the value, not the full buffer, which may have extra 0's
-                            seahash::hash(value) as i64,
-                            "invalid upsert state: checksum_sum does not match, state: {}",
-                            snapshotting
-                        );
-                        *self = Self::Value(Value::Value(
+                        if consolidating.checksum_sum.0 != seahash::hash(value) as i64 {
```

Comment:
Why did we downgrade the assert to an error here? Something needs to go very very wrong for the condition to be false so I would leave it as an assert here. Unless it is now more probable?

Reply:
I changed these to errors so that the callsite can log the error with context, like `source_id`, `worker_id`, etc. This was helpful in debugging, when I had these asserts failing and wouldn't know where and why they were failing. I can turn these back into asserts now but thought this might become handy in the future again.

Reply:
It's this commit that does those changes: https://github.com/MaterializeInc/materialize/pull/30499/commits/50ffa0ea83fea355a55093a540634f15163717b0, I isolated it because I thought we might want to roll that back